x/stake/types.go

```diff
@@ -57,23 +57,27 @@ const (
 // exchange rate. Voting power can be calculated as total bonds multiplied by
 // exchange rate.
 type Candidate struct {
-	Status      CandidateStatus `json:"status"`      // Bonded status
-	Address     sdk.Address     `json:"owner"`       // Sender of BondTx - UnbondTx returns here
-	PubKey      crypto.PubKey   `json:"pub_key"`     // Pubkey of candidate
-	Assets      sdk.Rat         `json:"assets"`      // total shares of a global hold pools
-	Liabilities sdk.Rat         `json:"liabilities"` // total shares issued to a candidate's delegators
-	Description Description     `json:"description"` // Description terms for the candidate
+	Status           CandidateStatus `json:"status"`            // Bonded status
+	Address          sdk.Address     `json:"owner"`             // Sender of BondTx - UnbondTx returns here
+	PubKey           crypto.PubKey   `json:"pub_key"`           // Pubkey of candidate
+	Assets           sdk.Rat         `json:"assets"`            // total shares of a global hold pools
+	Liabilities      sdk.Rat         `json:"liabilities"`       // total shares issued to a candidate's delegators
+	Description      Description     `json:"description"`       // Description terms for the candidate
+	ValidatorHeight  int64           `json:"validator_height"`  // Earliest height at current voting power
+	ValidatorCounter int64           `json:"validator_counter"` // Block-local tx index of validator change
```

Comment:
Hmm I almost want to say these names should be slightly more explicit.. like HeightAtBond / CountAtBond
Tokens are only ever considered to be "bonded" if they are in a part of a validator... open though.. maybe even more explicit like `ValidatorBondHeight` or `ValidatorBondCount`

Reply:
Also I imagine that the counter doesn't need to be an int64... how big is this ever going to get? I'd imagine int16 should be sufficient....... does the counter go up every tx in the block?

Reply:
LASTLY we should figure out a way for the counter to not be saved within the candidate struct as after the block is finished this term become irrelevant correct? - seems like an unnecessary field to hold... I'm open to keeping it here if you believe that taking it out will make the code considerably less maintainable, intiutively I think there's probably a fine solution where it doesn't need to be a candidate field however

Reply:
Renamed to `ValidatorBondHeight` and `ValidatorBondCounter`.

int16 should be sufficient, updated. Only incremented when necessary (when a candidate previously not a validator becomes a validator).

I also dislike storing the counter within the candidate. We need to retrieve the counter to know at which key the old validator can be accessed - https://github.com/cosmos/cosmos-sdk/blob/cwgoes/staking-validator-updates/x/stake/keeper.go#L134 - is there another way we could do that?

Reply:
I think let's table it for now - I think we may be able to wrap the validators by a meta-struct which just tacks on the counter on top - but that feels a bit clumsy / less maintainable / hard to understand if you don't already have a clear picture as to whats going on... maybe we could add some kind of a reference to the transaction-id / transaction-order instead of maintaining our own counter... I'll open up an issue - I think this is a post-launch kinda thing though

Reply:
https://github.com/cosmos/cosmos-sdk/issues/882